### ISSUE TITLE
Error: chol(): decomposition failed

### ISSUE THREAD

**Issue by amaa11** [NorskRegnesentral/shapr#329]
Dear all,

Thanks for such amazing tool.
I am facing an issue and would like you guide me to solve it.
I am running a binary classification using 1800 observations and 9 predictors  as training and 200 as test.
Below are the codes:
`model <- xgboost(data = X_train_m, label = y_train_m, nround = 500, verbose = FALSE, objective = "binary:logistic")`
`explainer <- shapr(X_train_m, model)`
`p <- mean(y_train_m)`
Then when I run:
`explanation <- explain(X_test_m, approach = "empirical", explainer = explainer,    prediction_zero = p)`
I got the error below:
`Error: chol(): decomposition failed`
Any idea what might be the cause?
Thanks in advance for your help.

best

**Comment by martinju** [comment]
Hi

Glad you like our software.

Without seeing data, it's not possible to say, but my guess is that you have data with extremely large values such that some internal covariance estimation fails due to numerical instability. Could you try scaling your data to e.g. standard deviation 1 before sending it to both xgboost and explain().

We are currently working towards a new release, and the plan is to implement internal scaling of the data such that issues like this does not occur. It is not implemented yet, however. In any case, feel free to play around with the development version of the package, available at branch devel.

**Comment by amaa11** [comment]
Thanks so much for your reply and suggestion. You are right this might be related to the data itself rather than the method/package. I just applied normalized both the training and the test as well using `scale` command, but still face same issue. 
I would love to share the data with you, but I am not allowed due to data protection issue. 
`X_train <- read.csv("X_train.csv")`
`y_train <- read.csv("y_train.csv")`
`X_test <- read.csv("X_test.csv")`
`y_test <- read.csv("y_test.csv")`
`X_train = scale(X_train)`
`X_test = scale(X_test)`
`y_test_m = as.matrix(y_test)`
`y_train_m = as.matrix(y_train)`
`model** <- xgboost(data = X_train, label = y_train_m, nround = 500, verbose = FALSE, objective = "binary:logistic")`
`explainer <- shapr(X_train, model)`
`p <- mean(y_train_m)`
`explanation <- explain(X_test, approach = "empirical", explainer = explainer,    prediction_zero = p)`
`Error: chol(): decomposition failed`

**Comment by martinju** [comment]
Without the data, it is hard for me to debug. My next suggestion would be to check for NAs in your data. That is not acceptable (we provide checks for that in the new shapr version) although xgboost can handle it. You would need to remove rows with NAs before sending them to shapr/explain.
If that is not it either, perhaps you have features in your data that are numerically identical?
If you could provide just a correlation matrix of your data, that would be helpful.

I assume the error comes from the mnvfast package that we use, so you may want to look here as well:
https://github.com/mfasiolo/mvnfast/issues/9

**Comment by amaa11** [comment]
I checked the NA values in both the training and test, both are FALSE.
Here is the correlation matrix. I think there is no identical features.
`      A           B        C             D        E           F            G       H        I`
`A  1.0000000  0.8768714 0.88128314 -0.28710984  0.8024573  0.8826721  0.7789518 0.79102028 -0.23436294`
`B  0.8768714  1.0000000 0.54558676 -0.69463273  0.7368539  0.7379084  0.7728540 0.53542546 -0.45233625`
`C   0.8812831  0.5455868 1.00000000  0.18235349  0.6745519  0.8133548  0.5981838 0.85263289  0.03611233`
`D  -0.2871098 -0.6946327 0.18235349  1.00000000 -0.2976615 -0.2069061 -0.4230079 0.06330774  0.58084386`
`E   0.8024573  0.7368539 0.67455192 -0.29766150  1.0000000  0.7568138  0.7083192 0.63639546 -0.28762719`
`F  0.8826721  0.7379084 0.81335476 -0.20690607  0.7568138  1.0000000  0.8931474 0.88514258 -0.28169294`
`G  0.7789518  0.7728540 0.59818383 -0.42300787  0.7083192  0.8931474  1.0000000 0.58127856 -0.67121517`
`H   0.7910203  0.5354255 0.85263289  0.06330774  0.6363955  0.8851426  0.5812786 1.00000000  0.18479670`
`I  -0.2343629 -0.4523363 0.03611233  0.58084386 -0.2876272 -0.2816929 -0.6712152 0.18479670  1.00000000`

**Comment by martinju** [comment]
OK, then I would suggest trying to remove one feature at a time, and possibly also one test observation/training observation at a time to see if you can narrow down which features/observations that cause the problem. 

My guess is that "something" happens when you are trying to invert a submatrix of the covariance matrix.

**Comment by amaa11** [comment]
Thanks so much for your guide.
I will do what you suggested and see if this will fix it or not.

**Comment by amaa11** [comment]
I have removed one feature each time and then run the model and the explainer as well.  I also removed 100 hundred of observations each time from the training and test. Finally the number of observations decreased from 1600 to 200 (in training) and in the test from 400 to 100. But the error is still there. Just as a note, I run five classifiers using these data in Python and then applied SHAP, and did not face this issue.

**Comment by martinju** [comment]
So the issue is still there with only two features, regardless of which features? That's very odd. 
Reduce the number of test observations to 2 and the number of training observations to say 100.
If this works you may start to add observations and/or features to see what features/observation causes the issue.

**Comment by amaa11** [comment]
Hi,

Thanks so much, after removing almost all features and adding only one each time, it worked with 2 features and all observations. 
As the title of the paper states, it is for individual prediction, so the method will be able to provide (local explanation) the contribution of each feature for a single or group of subjects, is it possible to provide the global explanation when the output indicates the overall contribution of each feature for all subjects?

**Comment by martinju** [comment]
Ok, I would then work iteratively to figure out which feature(s) is creating the issue, and explore further from that point.

For global explanation a common brute force approach is to take the mean of the absolute values of the individual shapley values per feature. However, this metric has not been properly studied, and does not possess the nice theoretical properties that the shapley values have. An alternative is to use SAGE: https://iancovert.com/blog/understanding-shap-sage/
Note, however, that the implementation of SAGE does not handle dependent features properly.

**Comment by amaa11** [comment]
Dear Martinju,

That was really helpful. I will look to the suggestions you mentioned and see if this will serve what I need.
Thanks for being patient with me to solve the issue.
Best

**Comment by martinju** [comment]
No, worries. Thanks
